Readme.md
# DIY project - SLIME SURVIVAL


## Make python game with: 

+ Python Arcade 2.6.17: https://api.arcade.academy/en/latest/index.html

+ Tiled : https://www.mapeditor.org/


## Resource:
+ Sprite: https://pipoya.itch.io/


## Gameplay:

+ kill slime --> collect items --> kill boss

ChangeLog.md
SLIME SURVIVAL 
===============	

--- v.0.x: basic gameplay ---
---------------

## v.0.4: player die
???

## v.0.3: player level up

### feat

- player's indexs
    + exp, level
    + healthy
    + atk
    + agility
    + def

- Player kill enemy
    + Deals damage to enemy
    + Receive exp when enemy die

- level up
    + level up --> increase exp to level up &  receive index

- add index
    + GUI add index
    + add index


## v.0.2: enemy movement & animation

### feat

- enemy: create --> follow --> attack --> die
    + create: has value exp
    + if around, follow player
    + if nearly, attack player
    + if its health is low, run away 
    + die
    + asset: sound: spawn, jump, attack, die

- Random monster spawn

- sound: add background sound

### refactor 

- split player to class Player

### fix

- is walking, die

## v.0.1: player movement & animation

### feat

- player animation: 4 directions top, down, left, right
    + Idle
    + Walk
    + Attack

- asset: sound walk, attack


## v.0.0: init - 10/09/2023

### feat
- Add map: center map (delta)
- Add sprite: player, slime

- Skeleton code:
    + Simple tile (map)
    + Player can move
    + Camera follow player

enemy.py
import math
import arcade
import random
from constain import *


class Enemy(arcade.Sprite):

    def __init__(self, name, start_x, start_y, damage=1, health=10):
        super().__init__()

        # index
        self.name = name
        self.action = "Idle"  # IDLE,WALK,ATTACK,DIE
        self.speed = ENEMY_SPEED
        self.step_full_health = 20
        self.step_run_away = 0
        self.damage = damage
        self.full_health = health
        self.health = health
        self.health_run_away = health//2
        self.target = None

        self.is_stuck = False
        self.texture_frame = 0  # 0--> last texture --> 0->...
        self.texture_frame_count = 0  # 0--> last texture --> 0->...

        self.init_x = start_x
        self.init_y = start_y
        self.center_x = start_x
        self.center_y = start_y
        self.hit_box_algorithm = 'Simple'

        # Load sound
        self.attack_sound = arcade.load_sound(SOUND_PATH+"slime_attack.mp3")
        self.walk_sound = arcade.load_sound(SOUND_PATH+"player_walk.wav")
        self.die_sound = arcade.load_sound(SOUND_PATH+"slime_die.wav")

        # Load textures
        self.idle_textures = get_texture_files("Plan_slime/Idle")
        self.walk_textures = get_texture_files("Plan_slime/Walk")
        self.attack_textures = get_texture_files("Plan_slime/Attack")
        self.die_textures = get_texture_files("Plan_slime/Die")

        self.scale = PLAYER_SCALING
        self.texture = self.idle_textures[0]

    def run_away(self):
        self.action = "Run_away"
        to_x = self.init_x
        to_y = self.init_y
        self.walk_to_taget(to_x , to_y)
        print(self.texture_frame)
        if self.texture_frame == self.step_full_health:
            self.health=self.full_health
            self.step_run_away=0

    def walk_around(self, start_x, start_y):
        self.action = "Walk_around"
        to_x = start_x + \
            random.randint(-ENEMY_VIEW_PADDING//2, ENEMY_VIEW_PADDING//2)
        to_y = start_y + \
            random.randint(-ENEMY_VIEW_PADDING//2, ENEMY_VIEW_PADDING//2)
        self.walk_to_taget(to_x, to_y)

    def auto_walking(self, end_x, end_y):
        # go around
        if self.action not in ['Walk', 'Die']:
            self.action = "Walk"
            self.texture_frame_count = 0
            self.texture_frame = 1
        self.walk_to_taget(end_x, end_y)
        if self.texture_frame % (len(self.walk_textures)*ENEMY_FPS) == 0:
            arcade.play_sound(self.walk_sound)
            self.action = "Idle"

    def walk_to_taget(self, target_x, target_y):
        x_diff = target_x - self.center_x
        y_diff = target_y - self.center_y
        self.center_x += self.change_x
        self.center_y += self.change_y
        angle = math.atan2(y_diff, x_diff)
        self.change_x = math.cos(angle) * self.speed
        self.change_y = math.sin(angle) * self.speed

    def is_attacking(self, target):
        if self.action not in ['Attack', 'Die']:
            arcade.play_sound(self.attack_sound)
            self.texture_frame_count = 0
            self.texture_frame = 1
            self.action = "Attack"
            target.health -= self.damage
        if (self.texture_frame == len(self.attack_textures)):
            self.is_idle()

    def is_idle(self):
        self.texture_frame_count = 0
        self.texture_frame = 1
        if self.action not in ['Die']:
            self.action = "Idle"

    def is_die(self):
        self.action = "Die"
        if (self.texture_frame == len(self.die_textures)):
            arcade.play_sound(self.die_sound)
            print("remove")
            self.remove_from_sprite_lists()

    def update_animation_and_sound(self):
        self.texture_frame_count += 1
        if self.texture_frame_count % ENEMY_FPS == 0:
            self.texture_frame += 1
            self.texture_frame_count = 0

        if self.action in ["Idle","Run_away","Walk_around"]:
            ind = self.texture_frame % (len(self.idle_textures))
            self.texture = self.idle_textures[ind]

        if self.action in ["Walk"]:
            ind = self.texture_frame % (len(self.walk_textures))
            self.texture = self.walk_textures[ind]

        if self.action in ["Attack"]:
            ind = self.texture_frame % (len(self.attack_textures))
            self.texture = self.attack_textures[ind]

        if self.action in ["Die"]:
            ind = self.texture_frame % (len(self.die_textures))
            self.texture = self.die_textures[ind]


player.py
import arcade

from constain import *


class Player(arcade.Sprite):
    MOVE_KEY = [
        arcade.key.LEFT,
        arcade.key.A,
        arcade.key.RIGHT,
        arcade.key.D,
        arcade.key.UP,
        arcade.key.W,
        arcade.key.DOWN,
        arcade.key.S
    ]
    ATTACK_KEY = [arcade.key.SPACE]

    def __init__(self, name, start_x, start_y, damage=2, health=10):
        super().__init__()

        self.name = name
        self.action = "Idle"  # IDLE,WALK,ATTACK,TELEPORTING
        self.face_direction = "DOWN"  # DOWN, LEFT, UP, RIGHT
        self.damage = damage
        self.health = health
        self.target = None

        self.texture_frame = 0  # 0--> last texture --> 0->...
        self.texture_frame_count = 0  # 0--> last texture --> 0->...

        self.center_x = start_x
        self.center_y = start_y
        self.hit_box_algorithm ='Simple'

        # Load sound
        self.attack_sound = arcade.load_sound(
            SOUND_PATH+"player_machine_hit.mp3")
        self.walk_sound = arcade.load_sound(SOUND_PATH+"player_walk.wav")
        self.die_sound = arcade.load_sound(SOUND_PATH+"player_die.wav")

        # Load textures
        self.idle_textures = {
            "UP": get_texture_files("Player_machine/Idle/UP"),
            "RIGHT": get_texture_files("Player_machine/Idle/RIGHT"),
            "DOWN": get_texture_files("Player_machine/Idle/DOWN"),
            "LEFT": get_texture_files("Player_machine/Idle/LEFT"),
        }
        self.walk_textures = {
            "UP": get_texture_files("Player_machine/Walk/UP"),
            "RIGHT": get_texture_files("Player_machine/Walk/RIGHT"),
            "DOWN": get_texture_files("Player_machine/Walk/DOWN"),
            "LEFT": get_texture_files("Player_machine/Walk/LEFT"),
        }
        self.attack_textures = {
            "UP": get_texture_files("Player_machine/Attack/UP"),
            "RIGHT": get_texture_files("Player_machine/Attack/RIGHT"),
            "DOWN": get_texture_files("Player_machine/Attack/DOWN"),
            "LEFT": get_texture_files("Player_machine/Attack/LEFT"),
        }
        self.scale = PLAYER_SCALING
        self.texture = self.idle_textures[self.face_direction][0]
        self.len_idle_textures = len(self.idle_textures[self.face_direction])
        self.len_walk_textures = len(self.walk_textures[self.face_direction])
        self.len_attack_textures = len(
            self.attack_textures[self.face_direction])

    def pan_camera_arround(self, camera, map_width, map_height):
        # This spot would center on the user
        if (self.center_x - camera.viewport_width / 2) < 0:
            camera_x = 0
        elif (self.center_x+camera.viewport_width / 2) > map_width:
            camera_x = map_width - camera.viewport_width
        else:
            camera_x = self.center_x - camera.viewport_width / 2
        if (self.center_y - camera.viewport_height / 2) < 0:
            camera_y = 0
        elif (self.center_y + camera.viewport_height / 2) > map_height:
            camera_y = map_height - camera.viewport_height
        else:
            camera_y = self.center_y - camera.viewport_height / 2
        camera.move_to([camera_x, camera_y])

    def on_key_press_action(self, key, modifiers):
        self.texture_frame_count = 0
        self.texture_frame = 1

        if key in Player.ATTACK_KEY:
            self.is_attacking()
        if key in Player.MOVE_KEY:
            self.change_face_direction(key)
            self.is_walking(PLAYER_SPEED)

    def change_face_direction(self, key):
        if key in [arcade.key.LEFT, arcade.key.A]:
            self.face_direction = "LEFT"
        if key in [arcade.key.RIGHT, arcade.key.D]:
            self.face_direction = "RIGHT"
        if key in [arcade.key.UP, arcade.key.W]:
            self.face_direction = "UP"
        if key in [arcade.key.DOWN, arcade.key.S]:
            self.face_direction = "DOWN"

    def on_key_release(self, key, modifiers):
        """
        Called when the user presses a mouse button.
        """
        if key in Player.MOVE_KEY:
            self.is_idle()
            self.change_x = 0
            self.change_y = 0

    def is_walking(self, speed):
        self.action = "Walk"
        if self.face_direction == "LEFT":
            self.change_x = - speed
        if self.face_direction == "RIGHT":
            self.change_x = speed
        if self.face_direction == "UP":
            self.change_y = speed
        if self.face_direction == "DOWN":
            self.change_y = - speed

    def is_attacking(self):
        self.action = "Attack"
        if (self.target is not None):
            self.target.health -= self.damage
            arcade.play_sound(self.attack_sound)

    def is_idle(self):
        self.action = "Idle"

    def update_animation_and_sound(self):
        self.texture_frame_count += 1
        if self.texture_frame_count % PLAYER_FPS.get(self.action, 1) == 0:
            self.texture_frame += 1
            self.texture_frame_count = 0
            if self.action == "Walk":
                arcade.play_sound(self.walk_sound)

        if self.action == "Idle":
            ind = self.texture_frame % (self.len_idle_textures)
            self.texture = self.idle_textures[self.face_direction][ind]

        if self.action == "Walk":
            ind = self.texture_frame % (self.len_walk_textures)
            self.texture = self.walk_textures[self.face_direction][ind]

        if self.action == "Attack":
            ind = self.texture_frame % (self.len_attack_textures)
            self.texture = self.attack_textures[self.face_direction][ind]
            if (self.texture_frame == self.len_attack_textures):
                self.target = None
                self.is_idle()


constain.py
import os
import arcade

SCREEN_TITLE = "Slime Survival"

# ---Sizing & Scaling
SCREEN_WIDTH = 1200
SCREEN_HEIGHT = 900

TILE_PIXEL_SIZE = 32
TILE_SCALING = 2

PLAYER_PIXEL_SIZE = 16
PLAYER_SCALING = 2
GRID_PIXEL_SIZE = TILE_PIXEL_SIZE * TILE_SCALING

# --- Margin
VIEWPORT_MARGIN_TOP = 60
VIEWPORT_MARGIN_BOTTOM = 60
VIEWPORT_RIGHT_MARGIN = 270
VIEWPORT_LEFT_MARGIN = 270

# --- Player
PLAYER_SPEED = 10
PLAYER_FPS = {
    "Idle": 30,
    "Walk": 15,
    "Attack": 15,
}


# ----- Enemy
ENEMY_FPS = 15
ENEMY_SPEED = 1
ENEMY_VIEW_PADDING = 5 * GRID_PIXEL_SIZE
ENEMY_DAME = 5
# --- Path

MAP_PATH = "./slime_survival/tile_world/"
MAP = {
    "delta": os.path.join(MAP_PATH, "map_delta.json")
}

SPRITE_PATH = "./slime_survival/sprite/"

SOUND_PATH = "./slime_survival/sound/"


# common function
def get_texture_files(name_folder):
    # Load textures for walking
    textures = []
    full_path = os.path.join(SPRITE_PATH, name_folder)+"/"
    if os.path.exists(full_path):
        all_file = os.listdir(full_path)
        for f in all_file:
            textures.append(arcade.load_texture(os.path.join(full_path, f)))
    return textures


main.py
import os
import math
import time
import arcade
from constain import *

from player import Player

from enemy import Enemy


class GameView(arcade.Window):
    """Main application class."""

    def __init__(self):
        """
        Initializer
        """
        super().__init__(SCREEN_WIDTH, SCREEN_HEIGHT, SCREEN_TITLE, resizable=True)

        # Tilemap Object
        self.tile_map = None
        self.map_width = SCREEN_WIDTH
        self.map_height = SCREEN_HEIGHT
        self.teleport = None

        # scene
        self.scene = None
        self.obstructions = []

        # player
        self.player = None
        self.player_engine = None

        # enemy
        self.enemy_list = None
        self.enemy_engine = None

        # Fps
        self.last_time = None
        self.frame_count = 0
        self.fps_message = None

        # Cameras
        self.camera = None
        self.gui_camera = None
        arcade.set_background_color(arcade.csscolor.CORNFLOWER_BLUE)

        # Load sounds
        self.teleport_sound = arcade.load_sound(SOUND_PATH+"teleport.wav")
        self.background_sound = arcade.load_sound(SOUND_PATH+"background.mp3")

    def setup(self):
        """Set up the game and initialize the variables."""
        self.old_action = ''

        # Scene
        self.scene = arcade.Scene()

        # Map
        self.create_map()
        self.scene = arcade.Scene.from_tilemap(self.tile_map)
        if self.tile_map.background_color:
            arcade.set_background_color(self.tile_map.background_color)
        self.obstructions = [self.tile_map.sprite_lists["fence_wall"],]

        # player
        self.scene.add_sprite_list_after("Player", "fence_wall")
        self.scene.add_sprite_list_before("Player", "tree")

        center_x = self.map_width / 2
        center_y = self.map_height / 2
        self.player = Player("Machine", center_x, center_y)
        self.scene.add_sprite("Player", self.player)

        self.player_engine = arcade.PhysicsEngineSimple(
            self.player,
            self.obstructions
        )

        # enemy
        self.scene.add_sprite_list_after("Enemy", "fence_wall")
        self.scene.add_sprite_list_before("Enemy", "tree")
        self.scene.add_sprite_list("Enemy")

        enemy = Enemy("Enemy_"+str(len(self.scene['Enemy'])),
                      center_x-TILE_PIXEL_SIZE*3, center_y-TILE_PIXEL_SIZE*3)
        self.scene.add_sprite("Enemy", enemy)
        self.enemy_engine = arcade.PhysicsEngineSimple(
            enemy,
            self.obstructions
        )

        # camera
        self.camera = arcade.Camera(self.width, self.height)
        self.gui_camera = arcade.Camera(self.width, self.height)

    def on_draw(self):
        """
        Render the screen.
        """
        self.clear()
        self.camera.use()
        self.scene.draw()
        self.gui_camera.use()
        self.calc_and_draw_fps()

        if (len(self.scene['Enemy']) > 0):
            arcade.draw_text(
                f"Enemy: {self.scene['Enemy'][0].action} {self.scene['Enemy'][0].step_run_away}",
                self.width-300, self.height-32,
                arcade.color.BLACK, 14)
            arcade.draw_text(
                f"{self.scene['Enemy'][0].health}",
                self.width-150, self.height-32*2,
                arcade.color.BLACK, 14)

        arcade.draw_text(
            f"Player: {self.player.action}",
            self.width-300, self.height-32*3,
            arcade.color.BLACK, 14)
        arcade.draw_text(
            f"{self.player.health}",
            self.width-150, self.height-32*4,
            arcade.color.BLACK, 14)

    def on_update(self, delta_time):
        """Movement and game logic"""

        # Call update on all sprites
        self.player_engine.update()
        self.camera.resize(self.width, self.height)
        self.gui_camera.resize(self.width, self.height)

        # player
        self.player.pan_camera_arround(
            self.camera, self.map_width, self.map_height)
        self.player.update_animation_and_sound()

        is_in_teleport = arcade.check_for_collision_with_lists(
            self.player, self.teleport,
        )
        for _ in is_in_teleport:
            if (self.player.action != "TELEPORTING"):
                self.player.action = "TELEPORTING"
                arcade.play_sound(self.teleport_sound)
        # enemy
        enemy_collision = arcade.check_for_collision_with_lists(
            self.player,
            [self.scene["Enemy"]]
        )

        if len(enemy_collision) > 0:
            self.player.target = enemy_collision[0]

        player_position = (self.player.center_x,
                           self.player.center_y)
        for enemy in self.scene['Enemy']:
            enemy.update_animation_and_sound()
            if enemy.health <= 0:
                enemy.is_die()
                continue
            if enemy.health <= enemy.health_run_away:
                enemy.run_away()
                continue
            if enemy in enemy_collision:
                enemy.is_attacking(self.player)
            else:
                is_sight = arcade.has_line_of_sight(player_position,
                                                    enemy.position,
                                                    self.scene['fence_wall'],
                                                    max_distance=ENEMY_VIEW_PADDING/2)
                if is_sight:
                    enemy.auto_walking(*player_position)
                else:
                    enemy.walk_around(enemy.init_x, enemy.init_y)

    def calc_and_draw_fps(self):
        # Start counting frames
        self.frame_count += 1
        # Calculate FPS if conditions are met
        if self.last_time and self.frame_count % 60 == 0:
            fps = 1.0 / (time.time() - self.last_time) * 60
            self.fps_message = f"FPS: {fps:5.0f}"
        # Draw FPS text

        if self.fps_message:
            arcade.draw_text(
                self.fps_message,
                10,
                40,
                arcade.color.BLACK,
                14
            )
        # Get time for every 60 frames
        if self.frame_count % 60 == 0:
            self.last_time = time.time()

    def on_key_press(self, key, modifiers):
        if key in self.player.MOVE_KEY or key in self.player.ATTACK_KEY:
            self.player.on_key_press_action(key, modifiers)

    def on_key_release(self, key, modifiers):
        self.player.on_key_release(key, modifiers)

    def create_map(self):
        #  maps
        layer_options = {
            "tree": {"use_spatial_hash": True},
            "fence_wall": {"use_spatial_hash": True},
            "background": {"use_spatial_hash": True},
            "ground": {"use_spatial_hash": True},
            "teleport": {"use_spatial_hash": True},
        }

        # Read in the tiled map
        self.tile_map = arcade.load_tilemap(
            MAP['delta'], layer_options=layer_options,
            scaling=TILE_SCALING
        )
        self.map_width = self.tile_map.width * GRID_PIXEL_SIZE
        self.map_height = self.tile_map.height * GRID_PIXEL_SIZE

        self.teleport = [
            self.tile_map.sprite_lists["teleport"],
        ]


def main():
    game = GameView()
    game.setup()
    arcade.play_sound((game.background_sound),
                      volume=0.5, looping=True, speed=0.7)
    arcade.run()


if __name__ == "__main__":
    main()
